<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">
    <!-- ROS2 Control plugin for Gazebo-->
    <ros2_control name="GazeboSystem" type="system">

        <!-- Hardware interface for GZ-->
        <hardware>
            <plugin>gz_ros2_control/GazeboSimSystem</plugin>
        </hardware>

        <!-- All continous joints-->
        <joint name="left_front_wheel">
            <command_interface name="velocity"/>
            <state_interface name="velocity"></state_interface>
            <state_interface name="position"></state_interface>
        </joint>

        <joint name="right_front_wheel">
            <command_interface name="velocity"/>
            <state_interface name="velocity"></state_interface>
            <state_interface name="position"></state_interface>
        </joint>

        <joint name="left_rear_wheel">
            <command_interface name="velocity"/>
            <state_interface name="velocity"></state_interface>
            <state_interface name="position"></state_interface>
        </joint>

        <joint name="right_rear_wheel">
            <command_interface name="velocity"/>
            <state_interface name="velocity"></state_interface>
            <state_interface name="position"></state_interface>
        </joint>

        <!-- All prismatic joints -->
        <joint name="right_rear_linear">
            <command_interface name="position"/>
            <command_interface name="velocity"/>
            <command_interface name="effort"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

        <joint name="right_front_linear">
            <command_interface name="position"/>
            <command_interface name="velocity"/>
            <command_interface name="effort"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

        <joint name="left_rear_linear">
            <command_interface name="position"/>
            <command_interface name="velocity"/>
            <command_interface name="effort"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

        <joint name="left_front_linear">
            <command_interface name="position"/>
            <command_interface name="velocity"/>
            <command_interface name="effort"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

        <!-- IMU sensor-->
        
        <sensor name="imu_sensor">
            <param name="frame_id">imu_link</param>

            <state_interface name="orientation.x"/>
            <state_interface name="orientation.y"/>
            <state_interface name="orientation.z"/>
            <state_interface name="orientation.w"/>
            <state_interface name="angular_velocity.x"/>
            <state_interface name="angular_velocity.y"/>
            <state_interface name="angular_velocity.z"/>
            <state_interface name="linear_acceleration.x"/>
            <state_interface name="linear_acceleration.y"/>
            <state_interface name="linear_acceleration.z"/>
        </sensor> 

    </ros2_control>
    
   <!-- <ros2_control name="imu_broad" type="sensor">
        <hardware>
            <plugin>gazebo_ros2_control/GazeboSensorSystem</plugin>
        </hardware>
        <sensor name="imu_sensor">
                <param name="frame_id">imu_link</param>
                <state_interface name="orientation.x" />
                <state_interface name="orientation.y" />
                <state_interface name="orientation.z" />
                <state_interface name="orientation.w" />
                <state_interface name="angular_velocity.x" />
                <state_interface name="angular_velocity.y" />
                <state_interface name="angular_velocity.z" />
                <state_interface name="linear_acceleration.x" />
                <state_interface name="linear_acceleration.y" />
                <state_interface name="linear_acceleration.z" />
            </sensor>
        </ros2_control> -->
 
        <!-- GZ plugin -->
    <gazebo>
          <plugin name="gz_ros2_control::GazeboSimROS2ControlPlugin" filename="libgz_ros2_control-system.so">
            <parameters>$(find kridtbot)/config/my_controllers.yaml</parameters>
        </plugin>   
    </gazebo>
    
</robot>